main/lsp/LSPTypechecker.cc

```diff
@@ -62,6 +64,88 @@ vector<ast::ParsedFile> sortParsedFiles(const core::GlobalState &gs, ErrorReport
+                 WorkerPool &workers, LSPFileUpdates::FastPathFilesToTypecheckResult result) {
+    UnorderedMap<core::FileRef, core::FoundDefHashes> oldFoundHashesForFiles;
+    config->logger->debug("Added {} files that were not part of the edit to the update set", result.extraFiles.size());
+    auto toTypecheck = move(result.extraFiles);
+
+    for (auto [fref, idx] : result.changedFiles) {
+        if (config->opts.lspExperimentalFastPathEnabled && !result.changedSymbolNameHashes.empty()) {
+            // Only set oldFoundHashesForFiles if symbols actually changed
+            // Means that no-op edits (and thus calls to LSPTypechecker::retypecheck) don't blow away
+            // methods only to redefine them with different IDs.
+
+            // Okay to `move` here (steals component of getFileHash) because we're about to use
+            // replaceFile to clobber fref.data(gs) anyways.
+            oldFoundHashesForFiles.emplace(fref, move(fref.data(gs).getFileHash()->foundHashes));
+        }
+
+        gs.replaceFile(fref, updates.updatedFiles[idx]);
+        // If file doesn't have a typed: sigil, then we need to ensure it's typechecked using typed: false.
+        fref.data(gs).strictLevel = pipeline::decideStrictLevel(gs, fref, config->opts);
+
+        toTypecheck.emplace_back(fref);
+    }
+    fast_sort(toTypecheck);
+    return make_pair(toTypecheck, oldFoundHashesForFiles);
+}
+
+vector<ast::ParsedFile>
+runIncrementalResolver(core::GlobalState &gs, shared_ptr<const LSPConfiguration> config, LSPFileUpdates &updates,
+                       WorkerPool &workers, shared_ptr<core::ErrorFlusher> errorFlusher,
+                       vector<core::FileRef> &toTypecheck,
+                       UnorderedMap<core::FileRef, core::FoundDefHashes> &oldFoundHashesForFiles,
+                       LSPFileUpdates::FastPathFilesToTypecheckResult result) {
+    ENFORCE(updates.typecheckingPath == PathType::Fast ||
+            updates.typecheckingPath == PathType::SlowWithIncrementalResolver);
+    auto path =
+        updates.typecheckingPath == PathType::Fast ? "fast path" : "slow path with incremental namer and resolver";
+    config->logger->debug("Running {} over num_files={}", path, toTypecheck.size());
+    unique_ptr<ShowOperation> op;
+    if (toTypecheck.size() > config->opts.lspMaxFilesOnFastPath / 2) {
+        op = make_unique<ShowOperation>(*config, ShowOperation::Kind::FastPath);
+    }
+    ENFORCE(gs.errorQueue->isEmpty());
+    vector<ast::ParsedFile> updatedIndexed;
+    for (auto &f : toTypecheck) {
+        // TODO(jvilk): We don't need to re-index files that didn't change.
+        // (`updates` has already-indexed trees, but they've been indexed with initialGS, not the
+        // `*gs` that we'll be typechecking with. We could do an ast::Substitute here if we had
+        // access to `initialGS`, but that's owned by the indexer thread, not this thread.)
+        auto t = pipeline::indexOne(config->opts, gs, f);
```

Comment:
This runs `index` too, not just incremental resolver. I don't think we can factor out as much code as this PR currently does. If anything, I think we only want small helper functions.

The slow path does its own thing for indexing, and thus can't share the code that does indexing with the fast path and incremental slow path.

Reply:
Right, but we need this indexing for resolver to be incremental

> The slow path does its own thing for indexing, and thus can't share the code that does indexing with the fast path and incremental slow path.

Yep, but this indexing is guarded by if, so it will be executed only on incremental slow path. I can change name to add clarity to it
https://github.com/sorbet/sorbet/blob/b5052fb6357e1ff1e66338403f6086509cdaeed6/main/lsp/LSPTypechecker.cc#L473-L475